### ISSUE TITLE
Linear kernel variance optimized to 0 with VGP model

### ISSUE THREAD

**Issue by vicow** [GPflow/GPflow#40]
Hello there,

We are trying a very simple example and don't understand why the variance of a linear kernel is learnt to be 0. Here is what we do:

````
import GPflow
import numpy as np
import scipy as sp
import tensorflow as tf

# Simple training set
m = 2
n = 4
X = np.array([
[ 1.,  3.],
[ 2.,  4.],
[ 1.,  1.],
[ 2.,  1.]])
Y = np.array([1, 1, 0, 0])
Y = np.expand_dims(Y, axis=1)

# Build and train the model
kernel = GPflow.kernels.Linear(m)
probit = GPflow.likelihoods.Bernoulli()
model = GPflow.vgp.VGP(X=X, Y=Y, likelihood=probit, kern=kernel)
model.optimize()
# print model

# Predict a new data point
x_new = np.array([1.5, 5])
model.predict_y([x_new])
````

If we display the model, we see that `model.kern.linear.variance` is 0. Which leads to a predictive mean of 0.5 all the time.

Is that the expected behavior ?

**Comment by jameshensman** [comment]
Yep, that's exactly the expected behaviour. 

The reason is that the linear kernel only supports functions which pass through zero. None of the functions in your prior are able to split the training set, so the best option is to give everything probability 0.5.

You can get the behaviour you expect by doing
```python
kernel = GPflow.kernels.Linear(m) + GPflow.kernels.Bias(m)
```
which allows an offset on the functions.

**Comment by lucasmaystre** [comment]
Actually, the dataset is linearly separable:

![ermau9ws5rav1wajhladalshpmizigu6kkwzqoatalshpmizigu6l0p73qkfvpuc5uaaaaaelftksuqmcc](https://cloud.githubusercontent.com/assets/278373/14522184/ade8c898-022d-11e6-9cd8-cc6f82017aeb.png)

Even if it wasn't separable, it is not intuitive to me that the "optimal" solution gives probability 0.5 to everything. Is it a consequence of taking the KL divergence from Q (the approximating distribution) instead of from P (the true posterior)?

**Comment by jameshensman** [comment]
The dataset is not separable by a function that also passes through (0, 0).
All functions in the Linear kernel pass through (0, 0).

No, this is a consequence of fitting the parameters by (approximate)
maximum likelihood. There is no parameter setting that can help the model
fit, to there's a mode in the parameter likelihood at the simplest model,
i.e. with zero variance.

Hope that helps.

**Comment by jameshensman** [comment]
NB. By function that pass through zero, I mean the Gaussian Process function, not the decision boundary (which you've plotted, and does pass through (0, 0)). The decision boundary made of the points where the GP function is equal to zero.

**Comment by lucasmaystre** [comment]
Here's how I see it:

1. The GP is a function f: R^2 -> R.
2. The linear kernel constrains the function to be a plane passing through the origin

Any plane whose intersection with the (x1, x2, 0) plane is the blue line / decision boundary in my plot above is a valid GP function with a linear kernel.

In particular, take the function: f(x1, x2) = -3*x1 + 2*x2. It seems to me that it is a valid function for a GP with a linear kernel. In particular, f(0, 0) = 0.

Am I still missing something?

**Comment by alexggmatthews** [comment]
Hi everyone,
Out of interest what happens in your original code if you substitute the following kernel?

kernel = GPflow.kernels.Linear(m) + GPflow.kernels.White(m)

Thanks,
Alex

**Comment by lucasmaystre** [comment]
@alexggmatthews adding the white noise kernel does not change things much. Output of `model.predict_y([x_new])` is:

    [array([[ 0.5]]), array([[ 0.25]])]

**Comment by markvdw** [comment]
Wouldn't this issue be easily fixed by adding either a Constant mean function or a Constant covariance function? Setting up the model like this:

```
mean = GPflow.mean_functions.Constant()
kernel = GPflow.kernels.Linear(m)
probit = GPflow.likelihoods.Bernoulli()
model = GPflow.vgp.VGP(X=X, Y=Y, likelihood=probit, kern=kernel, mean_function=mean)
```

Leads to a model which predicts 0.94 (as expected), if you terminate the optimisation before it goes haywire. The issue with this code is that the optimisation goes crazy due to doing maximum likelihood with a linearly separable dataset (variance of the linear kernel goes to infinity).

The fix would be to add a Constant covariance function, which gives Bayesian inference over a constant mean function for free.

**Comment by alexggmatthews** [comment]
@markvdw That sounds very feasible to me.

**Comment by lucasmaystre** [comment]
Thank you all for your input. The thing is, I'd like to really understand what is going on with the linear kernel - I am not (yet) convinced that setting the variance to 0 maximizes the marginal likelilhood.

I'll think a bit more about it next week and post back.

**Comment by markvdw** [comment]
Ok, so I think it's a bit different than what's said so far. Like @lucasmaystre and @vicow pointed out, the dataset is linearly separable using only the linear kernel. And, it is correct that the marginal likelihood is optimised by moving the variance to zero. Adding a constant mean/covariance function does "fix" the issue, in that it correctly classifies the points, but the marginal likelihood is actually poorer! This is is caused by a local optimum in the marginal likelihood. You can get a similar behaviour by initialising the Linear kernel variance to 10. It will also get stuck in a local optimum for the hyperparameters, but one that classifies correctly.

So what's going on? Well, the issue comes down to us optimising over the hyperparameters in the presence of very little data (4 points). Folk wisdom usually says that this makes even the hyperparameters susceptible to overfitting. However, a better description here is that using a point estimate for the hyperparameters is not a good reflection of the full posterior, and so it leads to odd results.

The marginal likelihood is a multivariate Gaussian evaluated at the data points. So the actual calculation contains a data fit term and a complexity penalty, which penalises placing probability mass on too many different datasets. To fit the dataset given, we need a linear latent function with a high slope, since the margin between the datasets is so small. In order to allow such functions under the posterior, we need to have a high variance for the linear kernel. However, in doing so, we're actually spreading the mass of the Gaussian over a larger area, making the compexity penalty larger. The funny thing here is that if we switch off the linear part of the kernel, we're essentially left with a model with all the data uncorrelated. This model fits the data less well, but also has a much smaller complexity penalty, and thus has a larger marginal likelihood.

This effect goes away completely simply by repeating the data twice, using this code:
```
X = np.repeat(X, 2, axis=0)
X += rnd.randn(*X.shape) * 0.1  # This line is not really necessary, just to prevent the kernel variance going to infinity sometimes
Y = np.repeat(Y, 2, axis=0)
```

It's good to note that if we were completely Bayesian about the hyperparameters as well, this effect would be less pronounced. The marginal likelihood isn't much bigger at the optimum than in the space around it, so if we were to run a sampler in the (approximate) hyperparameter posterior, we would predict using higher kernel variance a bit more, which would make the procedure at least classify the correct labels, even though the uncertainties would still be high. But actually, this would be the perfect behaviour in such a small-data example!

Code for demonstration:
```
# Build and train the model                                                               
mean = GPflow.mean_functions.Zero()
kernel = GPflow.kernels.Linear(m)
probit = GPflow.likelihoods.Bernoulli()
model = GPflow.vgp.VGP(X=X, Y=Y, likelihood=probit, kern=kernel, mean_function=mean)
model.optimize()
lml_orig = -model._objective(model.get_free_state())[0]

mean = GPflow.mean_functions.Zero()
kernel = GPflow.kernels.Linear(m)
probit = GPflow.likelihoods.Bernoulli()
model = GPflow.vgp.VGP(X=X, Y=Y, likelihood=probit, kern=kernel, mean_function=mean)
model.kern.variance = 10
model.optimize()
lml_init10 = -model._objective(model.get_free_state())[0]

mean = GPflow.mean_functions.Zero()
probit = GPflow.likelihoods.Bernoulli()
model = GPflow.vgp.VGP(X=X, Y=Y, likelihood=probit, kern=kernel, mean_function=mean)
model.kern = GPflow.kernels.Linear(m) + GPflow.kernels.Constant(m)
model.optimize()
lml_const = -model._objective(model.get_free_state())[0]

print("Linear init 1  %f" % lml_orig)
print("Linear init 10 %f" % lml_init10)
print("Linconst       %f" % lml_const)
```

**Comment by lucasmaystre** [comment]
@markvdw thank you very much - that makes it very clear! I never thought much about type 2 ML overfitting, that's a very nice example.

**Comment by vicow** [comment]
Thanks a lot @markvdw, very clear explanation :)
